### PR TITLE
[megatron] fix: correct batch_num_tokens aggregation for context parallelism

### LINKED ISSUE #5983: [megatron][engine][context parallel] batch_num_tokens aggregation undercounts valid tokens when CP > 1
### Summary

In the Megatron engine path, `batch_num_tokens` appears to be aggregated over the pure DP group in `transformer_impl.py`.

When `CP > 1`, each CP rank only holds a 1/CP slice of the sequence, so reducing `batch_num_tokens` over pure DP undercounts valid tokens by a factor of `CP`.

This affects `loss_agg_mode="token-mean"` normalization in the engine path.

### Current behavior

For `CP > 1`, `batch_num_tokens` is effectively smaller than the true valid-token count for the full DP×CP shard.

As a result, token-mean normalization is incorrect.

### Expected behavior

For `CP > 1`, `batch_num_tokens` should be reduced over the DP×CP group, while `dp_size` should remain pure DP.

### Proposed fix

Use `mpu.get_data_parallel_group(with_context_parallel=True)` when reducing `batch_num_tokens` in the Megatron engine path.

### Validation

I have:
- a minimal code fix in `transformer_impl.py`
- a focused test / reproducer for the CP>1 case
- manual smoke coverage for the engine path

### Related context

Related to #4852  
Potentially related to #1332

### Information

- [x] The official example scripts
- [ ] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [ ] My own task or dataset (give details below)

### Reproduction

The bug can be reproduced with a minimal Megatron engine-path setup where `context_parallel_size > 1`.

#### Config

- `tensor_model_parallel_size = 1`
- `pipeline_model_parallel_size = 1`
- `context_parallel_size = 2`
- `data_parallel_size = 1` (minimal case)
- `loss_agg_mode = "token-mean"`

#### Minimal reproducer

```python
import torch
import torch.distributed as dist
from megatron.core import parallel_state as mpu

def main():
    dist.init_process_group(backend="nccl")
    rank = dist.get_rank()
    local_rank = int(torch.cuda.current_device())
    device = torch.device("cuda", local_rank)

    # Assume Megatron parallel state has already been initialized with:
    # TP=1, PP=1, CP=2
    #
    # Each CP rank only sees a 1/CP slice of the sequence.
    # For a fully valid local slice:
    loss_mask = torch.ones(8, device=device)  # 8 valid tokens on each CP rank
    local_tokens = loss_mask.sum().to(torch.int64)  # = 8 on each rank

    # Current behavior in the buggy engine path:
    wrong = local_tokens.clone()
    dist.all_reduce(
        wrong,
        op=dist.ReduceOp.SUM,
        group=mpu.get_data_parallel_group(),   # pure DP group
    )

    # Expected behavior:
    correct = local_tokens.clone()
    dist.all_reduce(
        correct,
        op=dist.ReduceOp.SUM,
        group=mpu.get_data_parallel_group(with_context_parallel=True),  # DP×CP group
    )

    if rank == 0:
        print(f"local_tokens={int(local_tokens)}")
        print(f"wrong_batch_num_tokens={int(wrong)}")
        print(f"correct_batch_num_tokens={int(correct)}")

        # With CP=2, wrong_batch_num_tokens is smaller by a factor of 2.
        # This undercounts valid tokens and makes token-mean normalization incorrect.

if __name__ == "__main__":
    main()
```

### Expected behavior

For `CP > 1`, the Megatron engine path should aggregate `batch_num_tokens` over the correct DP×CP group, while keeping `dp_size` equal to pure DP. This should make `loss_agg_mode="token-mean"` use the correct valid-token count and avoid CP-dependent undercounting.

### PR DESCRIPTION
**Pull request by ianliuy**
### What does this PR do?

Fix `batch_num_tokens` undercounting when context parallelism (CP) > 1 in the Megatron engine path.

When `context_parallel_size > 1`, each CP rank holds 1/CP of the sequence. The `batch_num_tokens` all-reduce was using the pure DP group, missing other CP ranks' partial token counts — undercounting by a factor of CP. This causes incorrect `token-mean` loss normalization.

**Fix**: Use `mpu.get_data_parallel_group(with_context_parallel=True)` for both `batch_num_tokens` and `dp_size` in the static CP path, matching the pattern in `mtp_patch.py`. Dynamic context parallelism is guarded to preserve existing behavior.

Same class of fix as #5895 by @xhx1022. Fixes #5983. Related: #4852, #1332.

cc @wuxibin89 @vermouth1992 @xhx1022

### Checklist Before Starting

- [x] Search for similar PRs: [search link](https://github.com/verl-project/verl/pulls?q=is%3Apr+batch_num_tokens+context_parallel)
- [x] Format the PR title as `[{modules}] {type}: {description}`

### Test

Added CPU-only regression test (`tests/workers/test_batch_num_tokens_cp.py`) using `mp.spawn` with `gloo` backend:

- `test_batch_num_tokens_with_context_parallel`: DP=2, CP=2 — verifies pure DP group undercounts while DP+CP group gives correct total
- `test_batch_num_tokens_without_context_parallel`: DP=4, CP=1 — confirms no regression when CP=1

```
$ pytest tests/workers/test_batch_num_tokens_cp.py -v
tests/workers/test_batch_num_tokens_cp.py::test_batch_num_tokens_with_context_parallel PASSED
tests/workers/test_batch_num_tokens_cp.py::test_batch_num_tokens_without_context_parallel PASSED
```

### Design & Code Changes

**`verl/workers/engine/megatron/transformer_impl.py`** — `forward_backward_batch()`:

Before:
```python
batch_num_tokens = data["loss_mask"].sum().to(get_device_id())
torch.distributed.all_reduce(
    batch_num_tokens, op=torch.distributed.ReduceOp.SUM, group=self.get_data_parallel_group()
)
tu.assign_non_tensor(data, batch_num_tokens=batch_num_tokens.item())
tu.assign_non_tensor(data, dp_size=self.get_data_parallel_size())
```

After:
```python
if self.engine_config.dynamic_context_parallel:
    loss_reduce_group = self.get_data_parallel_group()
    dp_size = self.get_data_parallel_size()
else:
    loss_reduce_group = mpu.get_data_parallel_group(with_context_parallel=True)
    dp_size = mpu.get_data_parallel_world_size(with_context_parallel=True)
batch_num_tokens = data["loss_mask"].sum().to(get_device_id())
torch.distributed.all_reduce(
    batch_num_tokens, op=torch.distributed.ReduceOp.SUM, group=loss_reduce_group,
)
tu.assign_non_tensor(data, batch_num_tokens=batch_num_tokens.item())
tu.assign_non_tensor(data, dp_size=dp_size)
```

Key design decisions:
- **Dynamic CP guarded**: `dynamic_context_parallel` intentionally uses `dp_size=1` for batch distribution; preserved via explicit branch
- **Instance methods untouched**: `self.get_data_parallel_group()` is still used by `prepare_micro_batches` which correctly needs pure DP for micro-batch splitting
- **CP=1 safe**: `with_context_parallel=True` is a no-op when `context_parallel_size=1`

### Checklist Before Submitting

- [x] Read the [Contribute Guide](https://github.com/volcengine/verl/blob/main/CONTRIBUTING.md).
- [x] Apply pre-commit checks
- [x] Add unit test to cover the fix
- [ ] Once PR is ready for CI, send a message in the `ci-request` channel

### CODE REVIEW

File: verl/workers/engine/megatron/transformer_impl.py [hunk 1]
```diff
@@ -594,12 +594,25 @@ def forward_backward_batch(self, data: TensorDict, loss_function: Callable, forw
         tu.assign_non_tensor(data, sp_size=self.engine_config.context_parallel_size)
 
         # compute num_tokens in global batch for loss normalization
+        # When static context parallelism is enabled, each CP rank holds a portion of
```
Comment (vermouth1992) [inline verl/workers/engine/megatron/transformer_impl.py:597]: I think the data is not split into cp region here

**Review by vermouth1992 — CHANGES_REQUESTED** [review]

**Comment by wuxibin89** [comment]
https://github.com/verl-project/verl/pull/5993#issuecomment-4240925738
